File: server/shared/typescript/validation/default/src/default-email-validator.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import { TK } from "@d2/i18n-keys";
import { inputError, ok, validationFailed, type D2Result } from "@d2/result";
import { falsey } from "@d2/utilities";
import type { IEmailValidator } from "@d2/validation-abstractions";

/**
 * Email-format pattern — the cross-language source of truth shared with the
 * .NET `D2.Shared.Validation.DefaultEmailValidator.EMAIL_PATTERN` const. The
 * two literals are asserted byte-identical by a parity test, so any change
 * here MUST be mirrored on the .NET side.
 *
 * Enforces: total length 1-254, local part 1-64 of `A-Z 0-9 . _ % + -`,
 * domain labels of `A-Z 0-9 -` (no leading/trailing hyphen) separated by
 * dots with at least one dot. ASCII-only — the character classes reject any
 * non-ASCII codepoint.
 */
// long regex literal — cannot wrap
export const EMAIL_PATTERN =
  "^(?=.{1,254}$)[A-Z0-9._%+\\-]{1,64}@[A-Z0-9](?:[A-Z0-9\\-]{0,61}[A-Z0-9])?(?:\\.[A-Z0-9](?:[A-Z0-9\\-]{0,61}[A-Z0-9])?)+$";

// Case-insensitive ("i") to match the .NET RegexOptions.IgnoreCase. No "m"
// flag — `^`/`$` must anchor the whole string, not per-line.
const emailRegex = new RegExp(EMAIL_PATTERN, "i");

/**
 * Default `IEmailValidator` implementation. Mirrors the .NET
 * `D2.Shared.Validation.DefaultEmailValidator` — same pattern, same
 * normalization (trim then lowercase), same per-field `D2Result` contract.
 */
export class DefaultEmailValidator implements IEmailValidator {
  /**
   * Validates the supplied email and returns the normalized address on
   * success.
   *
   * @param email - The email address to validate (may be `undefined`,
   *   empty, or whitespace).
   * @returns `ok` wrapping the trimmed + lowercased address on success;
   *   `validationFailed` keyed `"email"` with `EMAIL_INVALID` on `undefined`,
   *   empty, whitespace, or structurally invalid input.
   */
  validate(email: string | undefined): D2Result<string> {
    if (falsey(email)) return DefaultEmailValidator.invalid();

    // `falsey` already excluded undefined / empty / whitespace.
    const trimmed = email!.trim();
    if (!emailRegex.test(trimmed)) return DefaultEmailValidator.invalid();

    return ok<string>(trimmed.toLowerCase());
  }

  private static invalid(): D2Result<string> {
    return validationFailed<string>({
      inputErrors: [inputError("email", [TK.common.validation.EMAIL_INVALID])],
    });
  }
}

File: public/packages/dotnet/validation/default/DefaultEmailValidator.cs
// -----------------------------------------------------------------------
// <copyright file="DefaultEmailValidator.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Validation;

using System.Text.RegularExpressions;
using DcsvIo.D2.I18n;
using DcsvIo.D2.Result;
using DcsvIo.D2.Utilities.Extensions;
using DcsvIo.D2.Validation.Abstractions;

/// <summary>
/// Default <see cref="IEmailValidator"/> — regex structural check enforcing
/// RFC-5321/5322 practical limits: total length 1–254, local part 1–64 of
/// ASCII alphanumeric plus <c>. _ % + -</c>, at-sign, one or more domain
/// labels separated by dots with no leading/trailing hyphen per label.
/// </summary>
/// <remarks>
/// <para>
/// Normalization: the input is trimmed then lowercased before the regex runs
/// and before the normalized value is returned. The comparison is therefore
/// case-insensitive by design; <c>RegexOptions.IgnoreCase</c> covers the
/// regex side.
/// </para>
/// <para>
/// <b>ReDoS posture: anchored, bounded — no catastrophic backtracking (Bucket B1).</b>
/// The pattern is anchored at both ends. The total-length lookahead is bounded. Every
/// character class is bounded: local part <c>{1,64}</c>, each domain label
/// <c>{0,61}</c>. The structure is linear in input length — genuinely B1
/// shape — but a 50 ms match timeout and the static-constructor JIT
/// pre-warm are retained as defense-in-depth against future-author regex edits that
/// could silently push the pattern into catastrophic-backtracking territory.
/// </para>
/// <para>
/// The pattern string is exposed as <see cref="EMAIL_PATTERN"/> so a parity
/// test can assert byte-identity with the TypeScript side.
/// </para>
/// </remarks>
public sealed class DefaultEmailValidator : IEmailValidator
{
    /// <summary>
    /// The email validation regex pattern — the cross-language source of truth
    /// shared with the TypeScript <c>@dcsv-io/d2-validation</c> package. Any change
    /// here must be mirrored on the TypeScript side; a parity test asserts
    /// byte-identity.
    /// </summary>
    // long regex literal — cannot wrap
    public const string EMAIL_PATTERN =
        @"^(?=.{1,254}$)[A-Z0-9._%+\-]{1,64}@[A-Z0-9](?:[A-Z0-9\-]{0,61}[A-Z0-9])?(?:\.[A-Z0-9](?:[A-Z0-9\-]{0,61}[A-Z0-9])?)+$";

    private static readonly TimeSpan sr_MatchTimeout = TimeSpan.FromMilliseconds(50);

    private static readonly Regex sr_Email = new(
        EMAIL_PATTERN,
        RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.CultureInvariant,
        sr_MatchTimeout);

    static DefaultEmailValidator()
    {
        // JIT pre-warm — defense-in-depth posture retains the timeout even for
        // a B1-shape pattern so that future edits to the pattern cannot
        // inadvertently remove the guard-rail.
        _ = sr_Email.IsMatch("a@b.co");
    }

    /// <inheritdoc />
    public D2Result<string> Validate(string? email)
    {
        if (email.Falsey())
            return Invalid();

        var trimmed = email!.Trim();

        bool isMatch;
        try
        {
            isMatch = sr_Email.IsMatch(trimmed);
        }
        catch (RegexMatchTimeoutException)
        {
            return Invalid();
        }

        if (!isMatch)
            return Invalid();

        return D2Result<string>.Ok(trimmed.ToLowerInvariant());
    }

    private static D2Result<string> Invalid()
        => D2Result<string>.ValidationFailed(
            inputErrors: [new InputError("email", [TK.Common.Validation.EMAIL_INVALID])]);
}
